File: src/tqi/scoring/reliability.py
"""Temporal reliability: travel time variability across departure times."""

import numpy as np

from tqi.scoring.tsr import valid_pair_mask

INF = float("inf")


def compute_reliability(
    mean_travel_time: np.ndarray,
    travel_time_std: np.ndarray,
    distances_km: np.ndarray,
) -> tuple[float, list[float]]:
    """Compute reliability metrics.

    Returns:
        mean_cv: Mean coefficient of variation across all valid, reachable pairs.
        per_origin_cv: List of mean CV per origin (for spatial mapping).
    """
    mask = valid_pair_mask(distances_km) & (mean_travel_time < INF) & (mean_travel_time > 0)

    # Coefficient of variation = std / mean
    with np.errstate(divide="ignore", invalid="ignore"):
        cv = np.where(mask, travel_time_std / mean_travel_time, np.nan)

    # Overall mean CV
    valid_cv = cv[mask]
    mean_cv = float(np.nanmean(valid_cv)) if len(valid_cv) > 0 else 0.0

    # Per-origin mean CV
    n_origins = mean_travel_time.shape[0]
    per_origin_cv = []
    for i in range(n_origins):
        row_mask = mask[i]
        if row_mask.any():
            per_origin_cv.append(float(np.nanmean(cv[i, row_mask])))
        else:
            per_origin_cv.append(0.0)

    return mean_cv, per_origin_cv

File: go/internal/scoring/reliability.go
package scoring

import "math"

// ComputeReliability returns the mean coefficient of variation (CV) across
// all valid reachable OD pairs and per-origin CVs.
func ComputeReliability(meanTT, ttStd, distKM [][]float64) (float64, []float64) {
	n := len(meanTT)
	perOrigin := make([]float64, n)

	var totalCV float64
	var totalCount int

	for i := 0; i < n; i++ {
		var originCV float64
		var originCount int

		for j := 0; j < len(meanTT[i]); j++ {
			if i == j {
				continue
			}
			if !IsValidPair(distKM[i][j]) {
				continue
			}
			if math.IsInf(meanTT[i][j], 1) || meanTT[i][j] >= InfFloat || meanTT[i][j] <= 0 {
				continue
			}
			cv := ttStd[i][j] / meanTT[i][j]
			originCV += cv
			originCount++
		}
		if originCount > 0 {
			perOrigin[i] = originCV / float64(originCount)
		}
		totalCV += originCV
		totalCount += originCount
	}

	var meanCV float64
	if totalCount > 0 {
		meanCV = totalCV / float64(totalCount)
	}
	return meanCV, perOrigin
}
